### PR TITLE
Add ut for etcd

### PR DESCRIPTION
**Pull request by Little-Wallace**
### What problem does this PR solve? 

Add some ut for worker crash because of some error and recover from config in etcd.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

File: dm/master/coordinator/coordinator.go [hunk 1]
```diff
@@ -189,8 +189,8 @@ func (c *Coordinator) HandleStartedWorker(w *Worker, cfg *config.MysqlConfig, su
 		c.taskConfigs[cfg.SourceID] = *cfg
 	} else {
 		w.SetStatus(WorkerFree)
-		delete(c.pendingtask, cfg.SourceID)
 	}
+	delete(c.pendingtask, cfg.SourceID)
```
Comment (csuzhangxc) [inline dm/master/coordinator/coordinator.go]: Is `pendingWorker` or another name with `Worker` more meaningful? the same with `taskConfigs`?

File: dm/master/coordinator/coordinator.go [hunk 2]
```diff
@@ -309,14 +313,21 @@ func (c *Coordinator) ObserveWorkers() {
...
-								// If this mysqltask has not been assigned to others, It could try to schedule on self.
+								// If the MySQL-task is still running in this worker.
 								c.schedule(source)
 								w.SetStatus(WorkerBound)
 								state = "bound"
+							} else if _, ok := c.upstreams[source]; !ok {
+								// If the MySQL-task has not been assigned to others, It could try to schedule on self.
+								c.upstreams[source] = w
+								w.SetStatus(WorkerBound)
+								c.schedule(source)
```
Comment (csuzhangxc) [inline dm/master/coordinator/coordinator.go]: Do we need to assign `state` here?

File: dm/worker/server.go [hunk 3]
```diff
@@ -140,7 +146,6 @@ func (s *Server) Start() error {
 
 	s.svr = grpc.NewServer()
 	pb.RegisterWorkerServer(s.svr, s)
-	fmt.Println("register worker")
```
Comment (csuzhangxc) [inline dm/worker/server.go]: How about removing two more usages in L90 and L236?

File: dm/master/coordinator/coordinator.go [hunk 4]
```diff
@@ -431,8 +431,14 @@ func (c *Coordinator) restartMysqlTask(w *Worker, cfg *config.MysqlConfig) bool
...
 		// remove sourceID from upstreams. So the source would be schedule in other worker.
 		delete(c.upstreams, cfg.SourceID)
+		delete(c.workerToConfigs, w.Address())
 		w.SetStatus(WorkerClosed)
 	}
 	c.mu.Unlock()
+	if w.State() == WorkerClosed {
+		ectx, cancel := context.WithTimeout(c.etcdCli.Ctx(), etcdTimeouit)
+		defer cancel()
+		c.etcdCli.Delete(ectx, common.UpstreamBoundWorkerKeyAdapter.Encode(w.Address()))
```
Comment (csuzhangxc) [inline dm/master/coordinator/coordinator.go]: Do we need to handle the returned `error` here?

File: dm/master/coordinator/coordinator.go (hunk 2 above)
Comment (Little-Wallace) [reply dm/master/coordinator/coordinator.go]: Yes

File: dm/master/coordinator/coordinator.go (hunk 4 above)
Comment (Little-Wallace) [reply dm/master/coordinator/coordinator.go]: Maybe an Error log?  It must receive a success response as long as the etcd cluster is still alive.

File: dm/master/coordinator/coordinator.go (hunk 1 above)
Comment (Little-Wallace) [reply dm/master/coordinator/coordinator.go]: How about 'pendingRequests' ?

File: dm/master/coordinator/coordinator.go (hunk 4 above)
Comment (csuzhangxc) [reply dm/master/coordinator/coordinator.go]: I think log an error is fine, it still can help us to debug.

File: dm/worker/server.go [hunk 5]
```diff
@@ -676,15 +676,22 @@ func (s *Server) OperateMysqlWorker(ctx context.Context, req *pb.MysqlWorkerRequ
...
-			resp.Msg = "Mysql task has not been created, please call CreateMysqlTask. Or there has been a worker started" +
-				" which has different config"
+			resp.Msg = errors.ErrorStack(err)
 			return resp, nil
 		}
 		w.Close()
+	} else if req.Op == pb.WorkerOp_StopWorker {
+		w, err := s.stopWorker(cfg.SourceID)
+		if err == terror.ErrWorkerSourceNotMatch {
+			resp.Result = false
```
Comment (csuzhangxc) [inline dm/worker/server.go]: How about setting `Msg` here too?

File: dm/ctl/common/util.go [hunk 6]
```diff
@@ -165,9 +165,9 @@ func GetFileContent(fpath string) ([]byte, error) {
 	return content, nil
 }
 
-// GetWorkerArgs extracts workers from cmd
-func GetWorkerArgs(cmd *cobra.Command) ([]string, error) {
-	return cmd.Flags().GetStringSlice("worker")
+// GetSourceArgs extracts workers from cmd
```
Comment (csuzhangxc) [inline dm/ctl/common/util.go]:
```suggestion
// GetSourceArgs extracts sources from cmd
```

File: dm/ctl/master/sql_replace.go [hunk 7]
```diff
@@ -53,21 +53,21 @@ func sqlReplaceFunc(cmd *cobra.Command, _ []string) {
...
-		if len(workers) != 0 {
-			common.PrintLines("--sharding operator always takes effect on DDL lock's owner, specified workers %v arguments will be ignored", workers)
+		if len(sources) != 0 {
+			common.PrintLines("--sharding operator always takes effect on DDL lock's owner, specified sources %v arguments will be ignored", sources)
 		}
 	} else {
-		if len(workers) != 1 {
-			common.PrintLines("should only specify one worker, but got %v", workers)
+		if len(sources) != 1 {
+			common.PrintLines("should only specify one worker, but got %v", sources)
```
Comment (csuzhangxc) [inline dm/ctl/master/sql_replace.go]:
```suggestion
			common.PrintLines("should only specify one source, but got %v", sources)
```

and 3 more `should only specify one worker` exist.

File: dm/proto/dmmaster.proto [hunk 8]
```diff
@@ -302,3 +305,13 @@ message RegisterWorkerResponse {
     bool result = 1;
     string msg = 2;
 }
+
+message OfflineWorkerRequest {
+    string name = 1;
+    string address = 2;
```
Comment (csuzhangxc) [inline dm/proto/dmmaster.proto]: 1. no _dmctl_ command added for `OfflineWorker`?
2. Is it safe for users to specify both of them (`name` and `address`) at the same time? Is it fine for only `name`?

Comment (Little-Wallace) [reply dm/proto/dmmaster.proto]: No. I check it and require the address and name of worker must match.

File: dm/ctl/master/sql_replace.go (hunk 7 above)
Comment (Little-Wallace) [reply dm/ctl/master/sql_replace.go]: Ok, I would refact them.

File: dm/ctl/common/util.go (hunk 6 above)
Comment (Little-Wallace) [reply dm/ctl/common/util.go]: I have fixed it.

File: dm/worker/server.go (hunk 5 above)
Comment (Little-Wallace) [reply dm/worker/server.go]: Yes. I made a mistake, and I have fixed it.

**Pull request merged**